### DISCUSSION TITLE
The normal equation N generated by writeNormalEquation is irreversible

### DISCUSSION THREAD

**Discussion by IamLittleMonster** [groops-devs/groops#176] [Q&A] [answered]
Dear developer,

Based on the foundation of #175 ,
1. After I executed GnssProcessing:processingStep:writeNormalEquation, four files were generated in the folder corresponding to outputfileNormalEquations, as mentioned in [doc1.7.17:](https://groops-devs.github.io/groops/html/fileFormat_normalEquation.html):
![image](https://user-images.githubusercontent.com/51938228/230297106-7912e857-837c-43c6-b78d-6097d2938fe6.png)
![image](https://user-images.githubusercontent.com/51938228/230297129-f1e2c52f-e588-431b-9880-260284abc0d5.png)

2. After converting normals.dat and normals.rightHandSide.dat into *. txt format, I opened the following results:
2.1 normals.txt:
![image](https://user-images.githubusercontent.com/51938228/230297764-2388a3a7-fc67-4ef3-9e71-3ec79853b73b.png)
2.2 normals.rightHandSide.txt:
![image](https://user-images.githubusercontent.com/51938228/230298192-78f1845e-e356-4da8-96cd-d1e9e7309217.png)

3. It is obvious that the result in 2.1. should not be the normal equation N, as it is irreversible. Therefore, I found the following theory from [Strass's 4.2.4](https://openlib.tugraz.at/reprocessing-multiple-gnss-constellations-and-a-global-station-network-from-1994-to-2020-with-the-raw-observation-approach-2022):
![image](https://user-images.githubusercontent.com/51938228/230299687-4650929a-8991-40f9-b696-4fa845b3f163.png)
Based on the above, I infer that the matrix in 2.1. should be W in the above formula. I am not sure if my inference is correct, and I hope the developer can answer.

4.If my inference in 3. is correct, according to formula (4.42) in 3, the calculated N is still irreversible, so I am not sure if there is a problem with my solution configuration. I hope the developer can help me.

5.When using [NormalsSolverVCE](https://groops-devs.github.io/groops/html/NormalsSolverVCE.html) for parameter calculation, according to the formula (4,41) in 3., we should need two parts: N and n. However, it seems that NormalsSolverVCE only allows the input of the normal equation N (perhaps W) and does not have the option to input rightHandSide. Therefore, I hope the developer can help me with this issue:
![image](https://user-images.githubusercontent.com/51938228/230301678-611cdf36-adcf-482f-aed7-79724d1eb6e9.png)

If the developer has enough time, I hope you can provide answers to the questions I raised in section 3.4.5.

With Kind Regards,
Hongzhan

**Comment by zhedumi** [comment, marked as answer]
Hi

Yes writeNormalEquation puts out four files. The normalequations, right handside, parameter names and info. These four files are regarded as a set of files throughout groops. 

On 2/3: No the matrix is not W. The cholesky decomposition is one way to solve the equation system of the normal equations without having the need of computing the inversion since computing the inversion would be a big issue in huge equation systems in term of memory and computaitonal power. One could also use other decompositions such as QR or something else.
The normal equations are symmetric and therefore are always only saved by a triangular part to save memory thats why the type on the header says UpperSymmetricMatrix. In that format with UpperSymmetricMatrix each column is a diagonal of the full matrix. The first column is the main diagonal, the second the first upper diagonal etc etc.
You can use matrixCalculate and use setType if you want to check it out and switch the matrix to full. 

On 5: You only need to link to the normals file itself the programm then will automatically look for the righthand file which is usually your normal filename with .rightHandSide appended.


Best regards,

Patrick

**Reply by IamLittleMonster** [reply]
Dear Patrick,

With your help, I have solved the transmitter antennaCenter.z (Z-direction PCO), but the above results are somewhat strange, which is very different from the Z-direction PCO in the official IGS antenna file.(**If the unit of the result is meters.**)
![image](https://user-images.githubusercontent.com/51938228/230324017-ec84cb35-c6eb-427e-931f-7072adae3800.png)
![image](https://user-images.githubusercontent.com/51938228/230324063-c4dbb22f-e4b5-46b5-8741-27e5522c6f39.png)

I noticed that in  [the POD example,](https://groops-devs.github.io/groops/html/cookbook.gnssNetwork.html) the **GnssProcessing: transmitter: inputfileAntennaDefinition** in 030groovsGnssProcessing. xml has already specified {groovsDataDir}/gnss/transmitter/antennaDefinition/igs/igs14/antennaDefinition in data. zip.  

Based on the above, I infer that the the transmitter antennaCenter.z (Z-direction PCO) I calculated is actually the corrected value for the transmitter antennaCenter.z (Z-direction PCO) in **GnssProcessing: transmitter: inputfileAntennaDefinition**. In other words, this is the offset of the Z-direction PCO.

May I ask the developer if my inference is correct, and if the unit of the result is meters?

**Reply by IamLittleMonster** [reply]
@zhedumi 
Dear Patrick,

I'm sorry to bother you, but I don't know your opinion on the above issue. I recalculated Galileo's satellite Z-direction PCO and also obtained a very small value. May I ask if it is related to the parameter selection of selectParameters in processingStep. Or it may be related to the relative order of selectParameterizations and estimate.

Best regards,

Hongzhan

**Reply by zhedumi** [reply]
Hi,
The inputfileAntennadefinition is the one given by IGS themself ([Check the files tab](https://igs.org/wg/antenna#files)) and for the different IGS reference frame work there is an inidividual antenna calibration. 
The small values could be the selection of the parameter types as described here [receiverAntennas](https://groops-devs.github.io/groops/html/gnssParametrizationType.html#receiverAntennas). Have not estimated any ACV or ACO for a long time so I currently do not have in mind what would be correct values or acceptable ones and sadly I will not have time to look into that until mid of may. One thing which I could be probably is to select the individual frequencies for each GNSS seperately for z-offsets or something similar. But as I said it has been ages since I last did any sort of ACO or ACV calculations so I am quiet out of touch on what the results should be in this regard with groops. 

Best regards,

Patrick

**Reply by IamLittleMonster** [reply]
Dear Patrick，

Thank you very much for your answer. Next, I will check my own problem from the source code.

Best regards,

Hongzhan
